File: server/connectors/api-connector/src/main/scala/com/prisma/api/connector/DataResolver.scala
package com.prisma.api.connector

import com.prisma.gc_values.{StringIdGCValue, IdGCValue}
import com.prisma.shared.models._

import scala.concurrent.Future

trait DataResolver {
  def project: Project

  def getModelForGlobalId(globalId: StringIdGCValue): Future[Option[Model]]

  def getNodeByWhere(where: NodeSelector, selectedFields: SelectedFields): Future[Option[PrismaNode]]

  def getNodes(model: Model, queryArguments: QueryArguments, selectedFields: SelectedFields): Future[ResolverResult[PrismaNode]]

  def getRelatedNodes(
      fromField: RelationField,
      fromNodeIds: Vector[IdGCValue],
      queryArguments: QueryArguments,
      selectedFields: SelectedFields
  ): Future[Vector[ResolverResult[PrismaNodeWithParent]]]

  def getScalarListValues(model: Model, listField: ScalarField, queryArguments: QueryArguments): Future[ResolverResult[ScalarListValues]]

  def getScalarListValuesByNodeIds(model: Model, listField: ScalarField, nodeIds: Vector[IdGCValue]): Future[Vector[ScalarListValues]]

  def getRelationNodes(relationTableName: String, queryArguments: QueryArguments): Future[ResolverResult[RelationNode]]

  def countByTable(table: String): Future[Int]

  def countByModel(model: Model, queryArguments: QueryArguments): Future[Int]
}

File: server/prisma-rs/query-engine/src/data_resolver.rs
mod sql;

use crate::node_selector::NodeSelector;
use crate::protobuf::prelude::*;
use prisma_common::PrismaResult;
use prisma_models::prelude::*;
use prisma_models::ScalarFieldRef;
pub use sql::*;

pub trait DataResolver {
    fn get_node_by_where(
        &self,
        node_selector: NodeSelector,
        selected_fields: SelectedFields,
    ) -> PrismaResult<Option<SingleNode>>;

    fn get_nodes(
        &self,
        model: ModelRef,
        query_arguments: QueryArguments,
        selected_fields: SelectedFields,
    ) -> PrismaResult<ManyNodes>;

    fn get_related_nodes(
        &self,
        from_field: RelationFieldRef,
        from_node_ids: Vec<GraphqlId>,
        query_arguments: QueryArguments,
        selected_fields: SelectedFields,
    ) -> PrismaResult<ManyNodes>;

    fn get_scalar_list_values_by_node_ids(
        &self,
        model: ModelRef,
        list_field: ScalarFieldRef,
        node_ids: Vec<GraphqlId>,
    ) -> PrismaResult<Vec<ScalarListValues>>;
}

pub struct ScalarListValues {
    pub node_id: GraphqlId,
    pub values: Vec<PrismaValue>,
}
